### DISCUSSION TITLE
iperf doesn't seem to work on servers with eBPF mounted

### DISCUSSION THREAD

**Discussion by hungryzzz** [esnet/iperf#1432] [Q&A]
# Context

* Version of iperf3:
server: iperf3.7
client: iperf3.9

* Hardware:
server: Intel(R) Core(TM) i5-9500T CPU @ 2.20GH
client: Intel(R) Core(TM) i7-10700F CPU @ 2.90GH

* Operating system (and distribution, if any):
server: Linux 5.15.0-53-generic #59~20.04.1-Ubuntu 
client: Linux  5.15.0-53-generic #59-Ubuntu 20.04.1

# Description
I don't know whether it is a bug...

* Expected Behavior
I load an XDP eBPF bytecodes in server, which changes the UDP packet's source and destination MAC, IP and port and sends back to the client. I've tested it manually, and it really works.

I use the cmd below in server and client respectively, want to test the throughput of the server with XDP eBPF:
```
# server
iperf3 -s -i 1 -p 8000

# client
iperf3 -u -c <server IP> -t 60 -p 8000 -l 512
```

* Actual Behavior
Both the client and server hang out, like below:
**server:**
```
-----------------------------------------------------------
Server listening on 8000
-----------------------------------------------------------
Accepted connection from <client IP>, port 55388
```

**client:**
```
Connecting to host <server IP>, port 8000
[  5] local <client IP> port 54111 connected to <server IP> port 8000
```

And I use tcpdump to trace the process, there is only two records:
```
00:55:08.055035 IP unicornt.54111 > <server IP>.8000: UDP, length 4
00:55:08.055333 IP <server IP>.8000 > unicornt.54111: UDP, length 4
```

* Steps to Reproduce
1. Compile the C code to eBPF bytecode using Clang, and load it in server (code is in attachment).
   - run the compile.sh
   - run the load.sh (update the network device name)
3. Start iperf3 in server.
4. Start iperf3 in client.
[ebpf.zip](https://github.com/esnet/iperf/files/10098809/ebpf.zip)

**Comment by swlars** [comment]
Can you say what your ultimate goal is? Modifying packets in the kernel might affect iperf3's dataflow.

**Reply by hungryzzz** [reply]
Yes, I want to use iperf3 to test and compare the throughput of server with and without eBPF mounted.

**Comment by davidBar-On** [comment]
Please provide more information by:
1. Run both server and client with the verbose & debug options `-V -d`.
2. Run tcpdump using the options `-X -xxx`.

Regarding:
> I load an XDP eBPF bytecodes in server, which changes the UDP packet's source and destination MAC, IP and port and sends back to the client.

I am not familiar with eBPF, so can you explain what is done (and why)?  Especially, is this only about messages sent from the server to the client?  If not, what is meant by "sent back to the client"?  This may be important, since when the server receive the first message from the client (connect), it deuces the clients IP address and port from the the received message.  From the information you sent, it seems that the 4 byte message that the server sent to the client (probably the 0x39383736 UDP_CONNECT_REPLY) was not received by iperf3.
